### DISCUSSION TITLE
Does terramate do change detection in remote modules?

### DISCUSSION THREAD

**Discussion by Tomasz-Kluczkowski** [terramate-io/terramate#1266] [Q&A] [answered]
In the change detection article here: https://terramate.io/docs/cli/change-detection/ I can only see change detection of local modules mentioned.

Does it mean that references to remote modules are not considered for change detection at all?

**Comment by i4ki** [comment]
Hi @Tomasz-Kluczkowski 

In the case of remote modules, their module version is pinned in the `module.source` then whenever you update the version, Terramate detects this change because the file was changed (normal change detection). AFAICS no special handling is needed for remote modules.
If you think that's not the case, can you elaborate?

**Comment by Tomasz-Kluczkowski** [comment]
Hi,

No I am at this stage considering if terramate could be a better tool for my company.

The issue I see is that we do not version modules independently - they are in a module repo which is tagged every time any module is updated for simplicity reasons.

I assume that this then means terramate would consider every update to a module tag as a change to all modules and that is not what I want to achieve.

Unless there is actual syntax/code analysis going on checking if given part actually changed?

Sent from Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by mariux** [reply, marked as answer]
Actually the answer would be "No, terramate will not detect all stacks as changed always when updating the remote repository"; Terramate just considers it a change if you change the tag in a reference to a module. When the module on the remote side changes, terramate is not aware of this. In case you do not reference anything or use refs like `main` or `latest`, terramate will actually not detect any remote changes ever.

If you reference a Terraform Module with something like `source = "github.com/org/repo?ref=v1.2"` and change the ref to `v1.3` then Terramate will notice the change in the change detection as the actual code changed. This is true for generated code or plain Terraform.

If you change the version in all stacks at once, then ofc Terramate will detect each changed stack and thus all.

When you want to do advanced versioning, then you can also use Globals inside the `source` attribute when referencing Terraform Modules in Code Generation e.g. `source = "github.com/org/vpc-module?ref=v${global.vpc_module_version}"`.

This will enable you to do things like `vpc_module_version = global.env == "prod" ? "1.2" : "1.3"` to update only environments that are not `"prod"` to `1.3` and keep `prod` stable. (given you have the environment stored in a `global.env` variable). In this case Change detection would only detect changes on environments not matching `"prod"`. 

Does this make the situation more clear?

**Reply by soerenmartius** [reply]
Hey @Tomasz-Kluczkowski , I'd be curious to hear your opinion about Terramate and if it fits your requirements. Feel free to join our [Discord](https://terramate.io/discord) and ping me there.
